src/peft/tuners/lora/eva.py

```diff
@@ -0,0 +1,673 @@
+
+
+# This is used to determine if inputs of two different layers are equal. For such cases, SVD
+# needs to be done for only for one of the equal inputs.
+class HashHook(_Hook):
+    """
+    A forward hook for hashing layer inputs. The hook is designed to be registered to a PyTorch module using the
+    `register_forward_hook` method.
+
+    This hook hashes the inputs of a specified layer during the forward pass of a neural network and stores the hash
+    values for later analysis or comparison.
+
+    Args:
+        name (str): Name of the layer to which this hook is attached. hashed_inputs (list): List of hashed inputs.
+        prepare_layer_inputs_fn (Optional[callable]): Function to prepare layer inputs for hashing.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        prepare_layer_inputs_fn: Optional[callable] = None,
+    ):
+        super().__init__(name, prepare_layer_inputs_fn)
+        self.hashed_inputs = []
+
+    @staticmethod
+    def hash_fn(tensor):
+        return hash(tuple(tensor.view(-1).tolist()))
+
+    @torch.no_grad()
+    def __call__(self, model, input, output):
+        x = self.prepare_layer_inputs(input).cpu()
+        self.hashed_inputs.append(self.hash_fn(x))
+
+
+def find_equal_values(dictionary: dict) -> dict:
+    """
+    Find keys in a dictionary that have the same value.
+
+    This function takes a dictionary and returns a new dictionary containing keys that have the same value. The keys in
+    the output dictionary are the values from the input dictionary, and the values are lists of keys that share the
+    same value.
+    """
+    value_dict = defaultdict(list)
+    for k, v in dictionary.items():
+        value_dict[v].append(k)
+    return {k: v for k, v in value_dict.items() if len(v) > 1}
+
+
+def get_target_name_key(name, pattern_keys):
+    """
+    Get the first substring of `name` that matches one of the keys in `pattern_keys`.
+    """
+    return next(filter(lambda key: re.match(rf".*\.{key}$", name), pattern_keys), name)
+
+
+def get_device_with_meta_params(model: torch.nn.Module) -> torch.device:
+    """
+    Get the device of the model's parameters. Useful if some parameters are on meta device.
+    """
+    devices = list({p.device for p in model.parameters() if p.device.type != "meta"})
+    if len(devices) > 1:
+        raise ValueError("model has multiple devices")
+    return devices[0]
+
+
+def move_inputs_to_device(inputs, device: Union[str, torch.device]):
+    """
+    Move the inputs to the specified device. Adapted from hf.Trainer.
+    """
+    if isinstance(inputs, Mapping):
+        return type(inputs)({k: move_inputs_to_device(v, device) for k, v in inputs.items()})
+    elif isinstance(inputs, (tuple, list)):
+        return type(inputs)(move_inputs_to_device(v, device) for v in inputs)
+    elif isinstance(inputs, torch.Tensor):
+        return inputs.to(device)
+    else:
+        raise ValueError(f"unsupported input type: {type(inputs)}")
```

Comment:
I wonder if we really need to raise here. Perhaps the  model somehow takes care of moving to the correct device. I think a warning is sufficient here.

Reply:
sounds good